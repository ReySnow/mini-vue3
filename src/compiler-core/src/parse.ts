import { NodeTypes } from "./ast"

const enum TagTypes {
    Start,
    End
}

export function baseParse(content) {
    const context = createParseContext(content)
    return createRoot(parseChildren(context, []))
}

function parseChildren(context, ancestors) {
    const nodes: any[] = []
    while (!isEnded(context, ancestors)) {
        let node
        let s = context.source
        if (s.startsWith('{{')) {
            node = paresInterpolation(context)
        } else if (s[0] === '<') {
            if (/[a-z]/i.test(s[1])) {
                node = parseElement(context, ancestors)
            }
        }

        if (!node) {
            node = parseText(context)
        }
        nodes.push(node)
    }
    return nodes
}

// 判断是否结束
function isEnded(context, ancestors) {
    let s = context.source
    if (s.startsWith('</')) {
        // 只要遇到结束标签就结束  从开始标签里面找
        for (let i = ancestors.length - 1; i >= 0; i--) {
            if (startsWithEndTagOpen(s, ancestors[i].tag)) {
                return true
            }
        }
    }
    // 没有需要解析的了
    return !s
}

function parseText(context: any): any {
    let endIndex = context.source.length
    let endTokens = ['{{', '<']// 解析text结束标识
    for (let item of endTokens) {
        let index = context.source.indexOf(item)
        if (index !== -1 && index < endIndex) {
            endIndex = index
        }
    }

    const content = parseTextData(context, endIndex)

    return {
        type: NodeTypes.TEXT,
        content
    }
}

function parseTextData(context, length) {
    const content = context.source.slice(0, length)
    advanceBy(context, length)
    return content
}

function parseElement(context: any, ancestors): any {
    const element: any = parseTag(context, TagTypes.Start)
    ancestors.push(element)
    element.children = parseChildren(context, ancestors)

    // 判断结束标签是否与开始标签一样
    if (startsWithEndTagOpen(context.source, element.tag)) {
        ancestors.pop()
        parseTag(context, TagTypes.End)
    } else {
        throw new Error(`缺少结束标签:${element.tag}`)
    }

    return element
}

function startsWithEndTagOpen(source, tag) {
    return source.startsWith('</')
        && source.slice(2, 2 + tag.length).toLowerCase() === tag.toLowerCase()
}

function parseTag(context, type) {
    const match: any = /^\<\/?([a-z]*)/i.exec(context.source)
    const tag = match[1]

    advanceBy(context, match[0].length)
    advanceBy(context, 1)// >右尖括号

    if (type === TagTypes.End) return

    return {
        type: NodeTypes.ELEMENT,
        tag
    }
}


function paresInterpolation(context) {
    const openDelimiter = '{{'
    const closeDelimiter = '}}'
    const closeIndex = context.source.indexOf(closeDelimiter, openDelimiter.length)

    // 去除开始部分
    advanceBy(context, openDelimiter.length)

    const rawContentLength = closeIndex - openDelimiter.length
    const rawContent = parseTextData(context, rawContentLength)
    const content = rawContent.trim()

    // 去除结束部分 }}
    advanceBy(context, closeDelimiter.length)

    return {
        type: NodeTypes.INTERPOLATION,
        content: {
            type: NodeTypes.SIMPLE_EXPRESSION,
            content: content,
        }
    }
}

// 剪切 向前推
function advanceBy(context, length) {
    context.source = context.source.slice(length)
}

function createRoot(children) {
    return {
        children,
        type: NodeTypes.ROOT
    }
}

function createParseContext(content: string) {
    return {
        source: content
    }
}
