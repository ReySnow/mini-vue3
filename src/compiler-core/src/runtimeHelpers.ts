
export const TO_DISPLAY_STRING = Symbol('toDisplayString')

export const CREAETE_ELEMENT_BLOCK = Symbol('createElementBlock')

export const helperMapNames = {
    [TO_DISPLAY_STRING]: 'toDisplayString',
    [CREAETE_ELEMENT_BLOCK]: 'createElementBlock'
}