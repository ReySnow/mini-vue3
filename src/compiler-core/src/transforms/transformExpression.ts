import { NodeTypes } from "../ast";

export function transformExpression(node) {
    if (node.type === NodeTypes.INTERPOLATION) {
        procesExpression(node.content)
    }
}

function procesExpression(node: any) {
    node.content = `_ctx.${node.content}`
}
