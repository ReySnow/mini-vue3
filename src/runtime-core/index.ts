export { h } from './h'
export { renderSlots } from './helpers/renderSlots'
export { createTextVNode } from './vnode'
export { getCurrentInstance } from './component'
export { provide, inject } from './apidInject'
export { createRenderer } from './render'
