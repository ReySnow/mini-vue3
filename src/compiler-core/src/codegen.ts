import { NodeTypes } from "./ast"
import { helperMapNames, TO_DISPLAY_STRING } from "./runtimeHelpers"

export function genreate(ast) {
    const context = createCodegenContext()
    const { push } = context

    genFunctionPreamble(ast, context)

    const functionName = 'render'
    const args = ['_ctx', '_catch']

    push(`function ${functionName}(${args.join(', ')}){`)
    push('return ')
    genNode(ast.codegenNode, context)
    push('}')
    return {
        code: context.code
    }
}

// 处理开头导入部分
function genFunctionPreamble(ast: any, context: any) {
    const { push } = context
    const VueBinging = 'Vue'
    // 别名
    const aliasHelper = s => `${helperMapNames[s]}: _${helperMapNames[s]}`
    if (ast.helpers.length > 0) {
        push(`const { ${ast.helpers.map(aliasHelper).join(', ')} } = ${VueBinging}`)
        push('\n')
    }
    push('return ')
}

// 处理节点
function genNode(node, context: any) {
    switch (node.type) {
        case NodeTypes.TEXT:
            genText(node, context)
            break;
        case NodeTypes.INTERPOLATION:
            genInterpolation(node, context)
            break;
        case NodeTypes.SIMPLE_EXPRESSION:
            genExpression(node, context)
            break
        default:
            break;
    }
}

// 处理插值中的表达式 此时已经被transform处理过了
function genExpression(node: any, context: any) {
    const { push } = context
    push(`${node.content}`)
}

// 处理插值
function genInterpolation(node: any, context: any) {
    const { push, helper } = context
    push(`${helper(TO_DISPLAY_STRING)}(`)
    genNode(node.content, context)
    push(')')
}

function genText(node: any, context: any) {
    const { push } = context
    push(`'${node.content}'`)
}

// 创建上下文对象
function createCodegenContext(): any {
    const context = {
        code: '',
        push(source) {
            context.code += source
        },
        helper(key) {
            return `_${helperMapNames[key]}`
        }
    }
    return context
}
