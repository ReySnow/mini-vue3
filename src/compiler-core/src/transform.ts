import { NodeTypes } from "./ast"
import { TO_DISPLAY_STRING } from "./runtimeHelpers"

export function transform(root, options = {}) {
    const context = createTransformContext(root, options)
    transformNode(root, context)
    creteRootCodegen(root)

    // 将节点需要的方法放到节点上， 在生成 code 的时候使用
    root.helpers = [...context.helpers.keys()]
}

function creteRootCodegen(root: any) {
    let child = root.children[0]
    if (child.type === NodeTypes.ELEMENT) {
        root.codegenNode = child.codegenNode
    } else {
        root.codegenNode = root.children[0]
    }
}

function transformNode(node: any, context) {
    // if (node.type === NodeTypes.TEXT) {
    //     node.content = node.content + ' mini-vue'
    // }
    // 通过外部传入的options实现对节点的处理，扩展性高
    // 进入时执行
    const nodeTransforms = context.nodeTransforms
    const exitFns: any = []
    for (let i = 0; i < nodeTransforms.length; i++) {
        const transform = nodeTransforms[i];
        const onExit = transform(node, context)
        if (onExit) {
            exitFns.push(onExit)
        }
    }

    // 处理节点
    switch (node.type) {
        case NodeTypes.INTERPOLATION:
            // 当为插值时添加插值对应的方法
            context.helper(TO_DISPLAY_STRING)
            break;
        case NodeTypes.ROOT:
        case NodeTypes.ELEMENT:
            traverseChildren(node, context)
            break
        default:
            break;
    }

    // 有的需要在退出时执行
    let i = exitFns.length
    while (i--) {
        exitFns[i]()
    }
}

function traverseChildren(node: any, context: any) {
    const children = node.children
    for (let item of children) {
        transformNode(item, context)
    }
}

function createTransformContext(root: any, options: any) {
    const context = {
        root,
        nodeTransforms: options?.nodeTransforms || [],
        helpers: new Map(),// 保存 xxx const {xxx}=Vue
        helper(key) {
            context.helpers.set(key, 1)
        }
    }
    return context;
}
