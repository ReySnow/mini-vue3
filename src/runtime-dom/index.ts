
import { createRenderer } from "../runtime-core";

const createElement = (type) => {
    return document.createElement(type)
}

const patchProp = (el, key, prevVal, nextVal) => {
    const isOn = (key: string) => /^on[A-Z]/.test(key)
    if (isOn(key)) {
        const even = key.slice(2).toLocaleLowerCase()
        el.addEventListener(even, nextVal)
    } else {
        if (nextVal === undefined || nextVal === null) {
            el.removeAttribute(key, nextVal)
        } else {
            el.setAttribute(key, nextVal)
        }
    }
}

const insert = (child, parent, anchor) => {
    // parent.append(el)
    parent.insertBefore(child, anchor || null)
}

// 移除一个元素
const remove = (child) => {
    const parent = child.parentNode
    if (parent) {
        parent.removeChild(child)
    }
}

const setElementText = (el, text) => {
    el.textContent = text
}

// 创建 runtime-dom 的自定义渲染器
const renderer: any = createRenderer({
    createElement,
    patchProp,
    insert,
    remove,
    setElementText
})

// 导出 runtime-dom 的createApp
export function createApp(...args) {
    return renderer.createApp(...args)
}

// runtime-dom -> runtime-core
export * from '../runtime-core'
