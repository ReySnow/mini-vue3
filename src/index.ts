export * from './runtime-dom'
import { baseCompiler } from './compiler-core/src'
/**
 * 依赖关系
 *     -> compiler
 * vue 
 *     -> runtime-dom -> runtime-core -> reactivity
 */


import * as runtimeDom from './runtime-dom'
import { registerRuntimeCompiler } from './runtime-dom'

function compileToFunction(template) {
    const { code } = baseCompiler(template)
    const render = new Function("Vue", code)(runtimeDom)
    return render
}

// 将 compiler 传入到 runtime-core component 中 
// runtime-core 不依赖于compiler-core
registerRuntimeCompiler(compileToFunction)
