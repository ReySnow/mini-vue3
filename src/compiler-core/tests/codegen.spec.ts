import { genreate } from "../src/codegen"
import { baseParse } from "../src/parse"
import { transform } from "../src/transform"
import { transformELement } from "../src/transforms/transformElement"
import { transformExpression } from "../src/transforms/transformExpression"
import { transformText } from "../src/transforms/transformText"

describe('codegen', () => {
    it('string', () => {
        const ast = baseParse('hi')
        transform(ast)
        const { code } = genreate(ast)
        expect(code).toMatchSnapshot()
    })

    it('should interpolation', () => {
        const ast = baseParse('{{message}}')
        transform(ast, {
            nodeTransforms: [transformExpression]
        })
        const { code } = genreate(ast)
        expect(code).toMatchSnapshot()
    })

    it('should element', () => {
        const ast = baseParse('<div>hi,{{message}}</div>')
        transform(ast, {
            // 执行顺序 transformExpression  transformText  transformELement
            nodeTransforms: [transformExpression, transformELement, transformText]
        })
        const { code } = genreate(ast)
        expect(code).toMatchSnapshot()
    })
})