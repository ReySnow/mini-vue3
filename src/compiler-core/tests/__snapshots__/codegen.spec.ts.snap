// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`codegen should element 1`] = `
"const { toDisplayString: _toDisplayString, createElementBlock: _createElementBlock } = Vue
return function render(_ctx, _catch){return _createElementBlock('div', null, 'hi,' + _toDisplayString(_ctx.message))}"
`;

exports[`codegen should interpolation 1`] = `
"const { toDisplayString: _toDisplayString } = Vue
return function render(_ctx, _catch){return _toDisplayString(_ctx.message)}"
`;

exports[`codegen string 1`] = `"return function render(_ctx, _catch){return 'hi'}"`;
